File: 프로그래머스/2/1844. 게임 맵 최단거리/게임 맵 최단거리.java
import java.util.*;

class Solution {
    static int[][] dirs = new int[][] {{1, 0}, {-1, 0}, {0, 1}, {0, -1}};
    public int solution(int[][] maps) {
        int answer = -1;
        int n = maps.length, m = maps[0].length;
        boolean[][] visited = new boolean[n][m];
        Queue<int[]> queue = new LinkedList<>();
        queue.add(new int[]{0, 0, 1});

        while (!queue.isEmpty()) {
            int[] pair = queue.poll();
            int i = pair[0], j = pair[1], cnt = pair[2];
            if (visited[i][j]) continue;
            visited[i][j] = true;

            if (i == n-1 && j == m-1) {
                answer = cnt;
                break;
            }

            for (int[] dir: dirs) {
                int di = dir[0], dj = dir[1];
                int ni = i + di, nj = j + dj;
                if (ni < 0 || nj < 0 || ni >= n || nj >= m) continue;
                if (maps[ni][nj] == 0 || visited[ni][nj]) continue;
                queue.add(new int[] {ni, nj, cnt + 1});
            }
        }

        return answer;
    }
}

File: 프로그래머스/2/1844. 게임 맵 최단거리/게임 맵 최단거리.py
import sys
from collections import deque
INVALID = -1

def solution(maps):
    n, m = len(maps), len(maps[0])
    # (1,1) -> (n,m)으로 bfs를 진행하고 거리를 반환. 갈 수 없으면 -1 반환
    visited = [[False for _ in range(m)] for _ in range(n)]
    distances = [[INVALID for _ in range(m)] for _ in range(n)]
    dxdys = ((0, 1), (0, -1), (1, 0), (-1, 0))
    
    queue = deque()
    queue.append((0, 0))
    visited[0][0] = True
    distances[0][0] = 1
    
    while queue:
        x, y = queue.popleft()
        for dx, dy in dxdys:
            nx, ny = x + dx, y + dy
            if nx < 0 or ny < 0 or nx >= n or ny >= m: continue
            if maps[nx][ny] == 0 or visited[nx][ny]: continue
            # 방문
            queue.append((nx, ny))
            visited[nx][ny] = True
            distances[nx][ny] = distances[x][y] + 1
    
    return distances[n-1][m-1]